File: app/src/main/java/com/msaggik/algorithmsdatastructures/java/graph/graph_weighted/DijkstrasAlgorithm.java
package com.msaggik.algorithmsdatastructures.java.graph.graph_weighted;

import com.msaggik.algorithmsdatastructures.java.graph.graph_weighted.entity.Road;
import com.msaggik.algorithmsdatastructures.java.graph.graph_weighted.entity.Vertex;

import java.util.PriorityQueue;

/**
 * Алгоритм Дейкстры
 * Упрощённая модель:  V - количество вершин, E - количество рёбер
 * применим к задачам:
 * - нахождение минимального расстояния между вершинами в взвешенном графе
 *
 * Вычислительная сложность: О(V^2 + E)
 **/

public class DijkstrasAlgorithm {

    public static void calculateDistance(Vertex startVertex) {
        startVertex.setDistance(0);
        PriorityQueue<Vertex> queue = new PriorityQueue<>();
        queue.add(startVertex); // добавление стартовой вершины в очередь

        while (!queue.isEmpty()) {
            Vertex vertex = queue.poll(); // извлечение вершины из очереди

            assert vertex != null;
            for (Road road: vertex.getRoads()) { // просмотр у неё всех дорог
                Vertex endVertex = road.getEndVertex(); // берётся одна из связующих вершин
                if (!endVertex.isVisited()) { // если данную вершину ещё не посещали

                    int distanceSum = vertex.getDistance() + road.getWeightRoad(); // берётся дистанция исходной точки и вес дороги

                    if (distanceSum < endVertex.getDistance()) { // если она меньше дистанции целевой точки (по умолчанию в ней она бесконечность)
                        endVertex.setDistance(distanceSum); // то устанавливаем в конечной точке данную дистанцию
                        queue.add(endVertex); // и добавляем конечную вершину в конец очереди
                        endVertex.setPredecessor(vertex); // ещё добавляем информацию о предыдущей вершине к конечной
                    }
                }
            }
            vertex.setVisited(true); // отмечаем что исходную вершину посетили
        }
    }

    public static void track(Vertex endVertex) {
        Vertex endPoint = endVertex;
        while (endPoint != null) {
            System.out.print(endPoint.getTitle() + " distance " + endPoint.getDistance() + ", ");
            endPoint = endPoint.getPredecessor();
        }
        System.out.print("\n");
    }
}

File: app/src/main/java/com/msaggik/algorithmsdatastructures/kotlin/graph/graph_weighted/DijkstrasAlgorithm.kt
package com.msaggik.algorithmsdatastructures.kotlin.graph.graph_weighted

import com.msaggik.algorithmsdatastructures.kotlin.graph.graph_weighted.entity.Vertex
import java.util.PriorityQueue

object  DijkstrasAlgorithm {

    fun calculateDistance(startVertex: Vertex) {
        startVertex.distance = 0
        val queue = PriorityQueue<Vertex>()
        queue.add(startVertex) // добавление стартовой вершины в очередь

        while (!queue.isEmpty()) {
            val vertex: Vertex = checkNotNull(queue.poll()) // извлечение вершины из очереди

            for (road in vertex.roads) { // просмотр у неё всех дорог
                val endVertex = road.endVertex// берётся одна из связующих вершин
                if (!endVertex.visited) { // если данную вершину ещё не посещали

                        val distanceSum = vertex.distance + road.weightRoad // берётся дистанция исходной точки и вес дороги

                        if (distanceSum < endVertex.distance) { // если она меньше дистанции целевой точки (по умолчанию в ней она бесконечность)
                            endVertex.distance = distanceSum // то устанавливаем в конечной точке данную дистанцию
                            queue.add(endVertex) // и добавляем конечную вершину в конец очереди
                            endVertex.predecessor = vertex // ещё добавляем информацию о предыдущей вершине к конечной
                        }
                    }
            }
            vertex.visited = true // отмечаем что исходную вершину посетили
        }
    }

    fun track(endVertex: Vertex?) {
        var endPoint = endVertex
        while (endPoint != null) {
            print(endPoint.title + " distance " + endPoint.distance + ", ")
            endPoint = endPoint.predecessor
        }
        print("\n")
    }
}
